Infrastructure for sending server parameters to clients, overriding settings on the client

Rogier-5 opened the pull request:
This set of commits allows selected options to be marked for sending to the client. They will be sent at client initialisation after a new connection has been established. The client will integrate them into its own settings such that they override values that the client may have set, or will set later. If the option is ever changed on the server, the new value will also be broadcast to all connected clients.

This feature can be used with client-side autopickup (#4427) for example, to allow the server to enforce autopickup privilege checking (on unmodified clients :-) See my branch [autopickup-with-client-config](https://github.com/Rogier-5/minetest/tree/autopickup-with-client-config) for the implementation.

Possible extensions:
- ~~Currently, the server-provided settings always override clients' settings. Another category of settings could have the status of suggestion, overriding default values on the client, but not the values the client has configured explicitly~~. (done)
- Lua interface to allow such settings to be created from mods (currently of little use; probably useful, when client-side scripting gets merged).

nerzhul commented on src/network/clientpackethandler.cpp:
```diff
@@ -764,6 +764,39 @@ void Client::handleCommand_ItemDef(NetworkPacket* pkt)
+		return;
+
+	// Decompress settings
+	std::string datastring(pkt->getString(2), pkt->getSize());
```
no need to serialize/compress for a little amount of data.
send u16 count + array[count] {std::string key, std::string value }

Rogier-5 replied:
Re: compression: In the short term, it won't be much indeed, but I assume that over time, the amount of data will grow, and even more so if this is going to be used for client-side scripting. And it may also end up being used for other purposes.
If preferred, I'll remove the compression.

Re: serialization: A value can be a nested group of settings, so the transmission format can't be just a list of `<key,value>` pairs. Also, I suppose a more generic serialization is more future-proof...

sfan5 replied:
Yes, please remove the compression.

rubenwardy replied:
A kilobyte on start up is pretty neglible, imo

Rogier-5 commented:
- Removed compression (can be added again later if desired).
- Fixed a remaining bug.
- Rebased to latest minetest tree
- Implemented extension nr. 1 above as well.

nerzhul commented on src/server.cpp:
```diff
@@ -1457,6 +1462,17 @@ void Server::Send(NetworkPacket* pkt)
 }
 
+
+void Server::Broadcast(NetworkPacket* pkt, int min_protocol_version, int max_protocol_version)
```
please use s32 instead of int

Rogier-5 replied:
OK

sfan5 replied:
Actuallly isn't the protocol version an `u8`?

Rogier-5 replied:
I think it's u16. I'll use whatever is used elsewhere.

nerzhul commented on src/settings.cpp:
```diff
@@ -1018,3 +1089,259 @@ void Settings::doCallbacks(const std::string &name) const
+
+
+// Must be entered with lock (m_mutex) held
+void Settings::deSerializeSettings(std::istream &is, SettingEntries *settings)
```
you should not use this sentinel, a count of messages in group or global list is sufficient

est31 commented:
:-1: stuff has worked without this infrastructure before, so I'm against it concept wise.

celeron55 commented:
How are vulnerabilities that can be potentially (or maybe already) caused by changing settings on the client prevented?

nerzhul commented:
@celeron55 changing GUI configuration, serverlist URL etc

nerzhul commented:
i think if this issue should be added, client should filter allowed keys to modify client settings, but i'm not for this, client side modding is more interesting, but not ready yet

lhofhansl commented:
Something like this would be useful to send - for example - max_block_send_distance to the client too. The client could then default to a reasonable fog distance (max_block_send_distance * 16), that way the fog would hide blocks rendered, and we'd get the optimal distance for this server.

Rogier-5 commented:
Security is a concern indeed.

I'll put settings that originate at the server in a separate namespace on both the client and the server. On the server, this will allow mods to add settings intended for the client (for client-side lua code). And it makes it absolutely clear in the client code that a server-originated setting is used, so thet they can't be used accidentally.

I'm thinking a separate group. E.g. on the server:
```c++
g_settings->set("client.some_setting", "some_value")
```
Possible code that could be used on the client:
```c++
myparam = g_settings->get("server.some_setting");
if (g_settings->get("server.some_setting") == "")
    myparam = g_settings->get("some_setting")
```
I assume this can be implemented using the existing setting groups feature.

Alternatively, a separate `Settings` instance could be used:
```C++
g_client_settings->set("some_setting", "some_value")
```
and
```c++
myparam = g_server_settings->get("some_setting");
```

rubenwardy commented:
:-1: for allowing servers to change client settings

However, a unified flag/parameter framework would be nice to make sending them easier. Eventually everything else could be ported over.

nerzhul commented:
i disagree too with this PR

paramat commented:
:-1:

paramat commented:
Disapproved and no progress.
